// n8n/dist/executions/executions.service.ee.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EEExecutionsService = void 0;
const WorkflowHelpers_1 = require("../WorkflowHelpers");
const executions_service_1 = require("./executions.service");
const workflows_services_ee_1 = require("../workflows/workflows.services.ee");
class EEExecutionsService extends executions_service_1.ExecutionsService {
    static async getWorkflowIdsForUser(user) {
        return (0, WorkflowHelpers_1.getSharedWorkflowIds)(user);
    }
    static async getExecution(req) {
        const execution = await super.getExecution(req);
        if (!execution)
            return;
        const relations = ['shared', 'shared.user', 'shared.role'];
        const workflow = (await workflows_services_ee_1.EEWorkflowsService.get({ id: execution.workflowId }, { relations }));
        if (!workflow)
            return;
        workflows_services_ee_1.EEWorkflowsService.addOwnerAndSharings(workflow);
        await workflows_services_ee_1.EEWorkflowsService.addCredentialsToWorkflow(workflow, req.user);
        execution.workflowData = {
            ...execution.workflowData,
            ownedBy: workflow.ownedBy,
            sharedWith: workflow.sharedWith,
            usedCredentials: workflow.usedCredentials,
        };
        return execution;
    }
}
exports.EEExecutionsService = EEExecutionsService;
