// n8n/dist/environments/sourceControl/sourceControl.controller.ee.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceControlController = void 0;
const decorators_1 = require("../../decorators");
const sourceControlEnabledMiddleware_ee_1 = require("./middleware/sourceControlEnabledMiddleware.ee");
const sourceControl_service_ee_1 = require("./sourceControl.service.ee");
const sourceControlPreferences_service_ee_1 = require("./sourceControlPreferences.service.ee");
const constants_1 = require("./constants");
const ResponseHelper_1 = require("../../ResponseHelper");
const express_1 = __importDefault(require("express"));
const typedi_1 = __importStar(require("typedi"));
const InternalHooks_1 = require("../../InternalHooks");
const sourceControlHelper_ee_1 = require("./sourceControlHelper.ee");
const sourceControlGetStatus_1 = require("./types/sourceControlGetStatus");
let SourceControlController = exports.SourceControlController = class SourceControlController {
    constructor(sourceControlService, sourceControlPreferencesService) {
        this.sourceControlService = sourceControlService;
        this.sourceControlPreferencesService = sourceControlPreferencesService;
    }
    async getPreferences() {
        return this.sourceControlPreferencesService.getPreferences();
    }
    async setPreferences(req) {
        var _a;
        if (req.body.branchReadOnly === undefined &&
            this.sourceControlPreferencesService.isSourceControlConnected()) {
            throw new ResponseHelper_1.BadRequestError('Cannot change preferences while connected to a source control provider. Please disconnect first.');
        }
        try {
            const sanitizedPreferences = {
                ...req.body,
                initRepo: (_a = req.body.initRepo) !== null && _a !== void 0 ? _a : true,
                connected: undefined,
                publicKey: undefined,
            };
            await this.sourceControlPreferencesService.validateSourceControlPreferences(sanitizedPreferences);
            const updatedPreferences = await this.sourceControlPreferencesService.setPreferences(sanitizedPreferences);
            if (sanitizedPreferences.initRepo === true) {
                try {
                    await this.sourceControlService.initializeRepository({
                        ...updatedPreferences,
                        branchName: updatedPreferences.branchName === ''
                            ? constants_1.SOURCE_CONTROL_DEFAULT_BRANCH
                            : updatedPreferences.branchName,
                        initRepo: true,
                    }, req.user);
                    if (this.sourceControlPreferencesService.getPreferences().branchName !== '') {
                        await this.sourceControlPreferencesService.setPreferences({
                            connected: true,
                        });
                    }
                }
                catch (error) {
                    await this.sourceControlService.disconnect({ keepKeyPair: true });
                    throw error;
                }
            }
            await this.sourceControlService.init();
            const resultingPreferences = this.sourceControlPreferencesService.getPreferences();
            void typedi_1.default.get(InternalHooks_1.InternalHooks).onSourceControlSettingsUpdated({
                branch_name: resultingPreferences.branchName,
                connected: resultingPreferences.connected,
                read_only_instance: resultingPreferences.branchReadOnly,
                repo_type: (0, sourceControlHelper_ee_1.getRepoType)(resultingPreferences.repositoryUrl),
            });
            return resultingPreferences;
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async updatePreferences(req) {
        try {
            const sanitizedPreferences = {
                ...req.body,
                initRepo: false,
                connected: undefined,
                publicKey: undefined,
                repositoryUrl: undefined,
            };
            const currentPreferences = this.sourceControlPreferencesService.getPreferences();
            await this.sourceControlPreferencesService.validateSourceControlPreferences(sanitizedPreferences);
            if (sanitizedPreferences.branchName &&
                sanitizedPreferences.branchName !== currentPreferences.branchName) {
                await this.sourceControlService.setBranch(sanitizedPreferences.branchName);
            }
            if (sanitizedPreferences.branchColor || sanitizedPreferences.branchReadOnly !== undefined) {
                await this.sourceControlPreferencesService.setPreferences({
                    branchColor: sanitizedPreferences.branchColor,
                    branchReadOnly: sanitizedPreferences.branchReadOnly,
                }, true);
            }
            await this.sourceControlService.init();
            const resultingPreferences = this.sourceControlPreferencesService.getPreferences();
            void typedi_1.default.get(InternalHooks_1.InternalHooks).onSourceControlSettingsUpdated({
                branch_name: resultingPreferences.branchName,
                connected: resultingPreferences.connected,
                read_only_instance: resultingPreferences.branchReadOnly,
                repo_type: (0, sourceControlHelper_ee_1.getRepoType)(resultingPreferences.repositoryUrl),
            });
            return resultingPreferences;
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async disconnect(req) {
        try {
            return await this.sourceControlService.disconnect(req.body);
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async getBranches() {
        try {
            return await this.sourceControlService.getBranches();
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async pushWorkfolder(req, res) {
        if (this.sourceControlPreferencesService.isBranchReadOnly()) {
            throw new ResponseHelper_1.BadRequestError('Cannot push onto read-only branch.');
        }
        try {
            await this.sourceControlService.setGitUserDetails(`${req.user.firstName} ${req.user.lastName}`, req.user.email);
            const result = await this.sourceControlService.pushWorkfolder(req.body);
            res.statusCode = result.statusCode;
            return result.statusResult;
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async pullWorkfolder(req, res) {
        try {
            const result = await this.sourceControlService.pullWorkfolder({
                force: req.body.force,
                variables: req.body.variables,
                userId: req.user.id,
            });
            res.statusCode = result.statusCode;
            return result.statusResult;
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async resetWorkfolder() {
        try {
            return await this.sourceControlService.resetWorkfolder();
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async getStatus(req) {
        try {
            const result = (await this.sourceControlService.getStatus(new sourceControlGetStatus_1.SourceControlGetStatus(req.query)));
            return result;
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async status(req) {
        try {
            return await this.sourceControlService.getStatus(new sourceControlGetStatus_1.SourceControlGetStatus(req.query));
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async generateKeyPair() {
        try {
            const result = await this.sourceControlPreferencesService.generateAndSaveKeyPair();
            return result;
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
};
__decorate([
    (0, decorators_1.Authorized)('none'),
    (0, decorators_1.Get)('/preferences', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "getPreferences", null);
__decorate([
    (0, decorators_1.Authorized)(['global', 'owner']),
    (0, decorators_1.Post)('/preferences', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "setPreferences", null);
__decorate([
    (0, decorators_1.Authorized)(['global', 'owner']),
    (0, decorators_1.Patch)('/preferences', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "updatePreferences", null);
__decorate([
    (0, decorators_1.Authorized)(['global', 'owner']),
    (0, decorators_1.Post)('/disconnect', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "disconnect", null);
__decorate([
    (0, decorators_1.Authorized)('any'),
    (0, decorators_1.Get)('/get-branches', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "getBranches", null);
__decorate([
    (0, decorators_1.Authorized)(['global', 'owner']),
    (0, decorators_1.Post)('/push-workfolder', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedAndEnabledMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "pushWorkfolder", null);
__decorate([
    (0, decorators_1.Authorized)(['global', 'owner']),
    (0, decorators_1.Post)('/pull-workfolder', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedAndEnabledMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object]),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "pullWorkfolder", null);
__decorate([
    (0, decorators_1.Authorized)(['global', 'owner']),
    (0, decorators_1.Get)('/reset-workfolder', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedAndEnabledMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "resetWorkfolder", null);
__decorate([
    (0, decorators_1.Authorized)('any'),
    (0, decorators_1.Get)('/get-status', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedAndEnabledMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "getStatus", null);
__decorate([
    (0, decorators_1.Authorized)('any'),
    (0, decorators_1.Get)('/status', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "status", null);
__decorate([
    (0, decorators_1.Authorized)(['global', 'owner']),
    (0, decorators_1.Post)('/generate-key-pair', { middlewares: [sourceControlEnabledMiddleware_ee_1.sourceControlLicensedMiddleware] }),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], SourceControlController.prototype, "generateKeyPair", null);
exports.SourceControlController = SourceControlController = __decorate([
    (0, typedi_1.Service)(),
    (0, decorators_1.RestController)(`/${constants_1.SOURCE_CONTROL_API_ROOT}`),
    __metadata("design:paramtypes", [sourceControl_service_ee_1.SourceControlService,
        sourceControlPreferences_service_ee_1.SourceControlPreferencesService])
], SourceControlController);
