// n8n/dist/environments/sourceControl/sourceControlExport.service.ee.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SourceControlExportService = void 0;
const typedi_1 = require("typedi");
const path_1 = __importDefault(require("path"));
const constants_1 = require("./constants");
const Db = __importStar(require("../../Db"));
const n8n_workflow_1 = require("n8n-workflow");
const promises_1 = require("fs/promises");
const fs_1 = require("fs");
const n8n_core_1 = require("n8n-core");
const sourceControlHelper_ee_1 = require("./sourceControlHelper.ee");
const typeorm_1 = require("typeorm");
const variables_service_1 = require("../variables/variables.service");
const repositories_1 = require("../../databases/repositories");
let SourceControlExportService = exports.SourceControlExportService = class SourceControlExportService {
    constructor(variablesService, tagRepository) {
        this.variablesService = variablesService;
        this.tagRepository = tagRepository;
        this.replaceCredentialData = (data) => {
            for (const [key] of Object.entries(data)) {
                try {
                    if (data[key] === null) {
                        delete data[key];
                    }
                    else if (typeof data[key] === 'object') {
                        data[key] = this.replaceCredentialData(data[key]);
                    }
                    else if (typeof data[key] === 'string') {
                        data[key] = (0, sourceControlHelper_ee_1.stringContainsExpression)(data[key]) ? data[key] : '';
                    }
                    else if (typeof data[key] === 'number') {
                        continue;
                    }
                }
                catch (error) {
                    n8n_workflow_1.LoggerProxy.error(`Failed to sanitize credential data: ${error.message}`);
                    throw error;
                }
            }
            return data;
        };
        const userFolder = n8n_core_1.UserSettings.getUserN8nFolderPath();
        this.gitFolder = path_1.default.join(userFolder, constants_1.SOURCE_CONTROL_GIT_FOLDER);
        this.workflowExportFolder = path_1.default.join(this.gitFolder, constants_1.SOURCE_CONTROL_WORKFLOW_EXPORT_FOLDER);
        this.credentialExportFolder = path_1.default.join(this.gitFolder, constants_1.SOURCE_CONTROL_CREDENTIAL_EXPORT_FOLDER);
    }
    getWorkflowPath(workflowId) {
        return (0, sourceControlHelper_ee_1.getWorkflowExportPath)(workflowId, this.workflowExportFolder);
    }
    getCredentialsPath(credentialsId) {
        return (0, sourceControlHelper_ee_1.getCredentialExportPath)(credentialsId, this.credentialExportFolder);
    }
    async deleteRepositoryFolder() {
        try {
            await (0, promises_1.rm)(this.gitFolder, { recursive: true });
        }
        catch (error) {
            n8n_workflow_1.LoggerProxy.error(`Failed to delete work folder: ${error.message}`);
        }
    }
    rmFilesFromExportFolder(filesToBeDeleted) {
        try {
            filesToBeDeleted.forEach((e) => (0, fs_1.rmSync)(e));
        }
        catch (error) {
            n8n_workflow_1.LoggerProxy.error(`Failed to delete workflows from work folder: ${error.message}`);
        }
        return filesToBeDeleted;
    }
    async writeExportableWorkflowsToExportFolder(workflowsToBeExported, owners) {
        await Promise.all(workflowsToBeExported.map(async (e) => {
            const fileName = this.getWorkflowPath(e.id);
            const sanitizedWorkflow = {
                id: e.id,
                name: e.name,
                nodes: e.nodes,
                connections: e.connections,
                settings: e.settings,
                triggerCount: e.triggerCount,
                versionId: e.versionId,
                owner: owners[e.id],
            };
            n8n_workflow_1.LoggerProxy.debug(`Writing workflow ${e.id} to ${fileName}`);
            return (0, promises_1.writeFile)(fileName, JSON.stringify(sanitizedWorkflow, null, 2));
        }));
    }
    async exportWorkflowsToWorkFolder(candidates) {
        try {
            (0, sourceControlHelper_ee_1.sourceControlFoldersExistCheck)([this.workflowExportFolder]);
            const workflowIds = candidates.map((e) => e.id);
            const sharedWorkflows = await Db.collections.SharedWorkflow.find({
                relations: ['role', 'user'],
                where: {
                    role: {
                        name: 'owner',
                        scope: 'workflow',
                    },
                    workflowId: (0, typeorm_1.In)(workflowIds),
                },
            });
            const workflows = await Db.collections.Workflow.find({
                where: {
                    id: (0, typeorm_1.In)(workflowIds),
                },
            });
            const owners = {};
            sharedWorkflows.forEach((e) => (owners[e.workflowId] = e.user.email));
            await this.writeExportableWorkflowsToExportFolder(workflows, owners);
            return {
                count: sharedWorkflows.length,
                folder: this.workflowExportFolder,
                files: workflows.map((e) => ({
                    id: e === null || e === void 0 ? void 0 : e.id,
                    name: this.getWorkflowPath(e === null || e === void 0 ? void 0 : e.name),
                })),
            };
        }
        catch (error) {
            throw Error(`Failed to export workflows to work folder: ${error.message}`);
        }
    }
    async exportVariablesToWorkFolder() {
        try {
            (0, sourceControlHelper_ee_1.sourceControlFoldersExistCheck)([this.gitFolder]);
            const variables = await this.variablesService.getAllCached();
            if (variables.length === 0) {
                return {
                    count: 0,
                    folder: this.gitFolder,
                    files: [],
                };
            }
            const fileName = (0, sourceControlHelper_ee_1.getVariablesPath)(this.gitFolder);
            const sanitizedVariables = variables.map((e) => ({ ...e, value: '' }));
            await (0, promises_1.writeFile)(fileName, JSON.stringify(sanitizedVariables, null, 2));
            return {
                count: sanitizedVariables.length,
                folder: this.gitFolder,
                files: [
                    {
                        id: '',
                        name: fileName,
                    },
                ],
            };
        }
        catch (error) {
            throw Error(`Failed to export variables to work folder: ${error.message}`);
        }
    }
    async exportTagsToWorkFolder() {
        try {
            (0, sourceControlHelper_ee_1.sourceControlFoldersExistCheck)([this.gitFolder]);
            const tags = await this.tagRepository.find();
            if (tags.length === 0) {
                return {
                    count: 0,
                    folder: this.gitFolder,
                    files: [],
                };
            }
            const mappings = await Db.collections.WorkflowTagMapping.find();
            const fileName = path_1.default.join(this.gitFolder, constants_1.SOURCE_CONTROL_TAGS_EXPORT_FILE);
            await (0, promises_1.writeFile)(fileName, JSON.stringify({
                tags: tags.map((tag) => ({ id: tag.id, name: tag.name })),
                mappings,
            }, null, 2));
            return {
                count: tags.length,
                folder: this.gitFolder,
                files: [
                    {
                        id: '',
                        name: fileName,
                    },
                ],
            };
        }
        catch (error) {
            throw Error(`Failed to export variables to work folder: ${error.message}`);
        }
    }
    async exportCredentialsToWorkFolder(candidates) {
        try {
            (0, sourceControlHelper_ee_1.sourceControlFoldersExistCheck)([this.credentialExportFolder]);
            const credentialIds = candidates.map((e) => e.id);
            const credentialsToBeExported = await Db.collections.SharedCredentials.find({
                relations: ['credentials', 'role', 'user'],
                where: {
                    credentialsId: (0, typeorm_1.In)(credentialIds),
                },
            });
            let missingIds = [];
            if (credentialsToBeExported.length !== credentialIds.length) {
                const foundCredentialIds = credentialsToBeExported.map((e) => e.credentialsId);
                missingIds = credentialIds.filter((remote) => foundCredentialIds.findIndex((local) => local === remote) === -1);
            }
            const encryptionKey = await n8n_core_1.UserSettings.getEncryptionKey();
            await Promise.all(credentialsToBeExported.map(async (sharedCredential) => {
                const { name, type, nodesAccess, data, id } = sharedCredential.credentials;
                const credentialObject = new n8n_core_1.Credentials({ id, name }, type, nodesAccess, data);
                const plainData = credentialObject.getData(encryptionKey);
                const sanitizedData = this.replaceCredentialData(plainData);
                const fileName = this.getCredentialsPath(sharedCredential.credentials.id);
                const sanitizedCredential = {
                    id: sharedCredential.credentials.id,
                    name: sharedCredential.credentials.name,
                    type: sharedCredential.credentials.type,
                    data: sanitizedData,
                    nodesAccess: sharedCredential.credentials.nodesAccess,
                };
                n8n_workflow_1.LoggerProxy.debug(`Writing credential ${sharedCredential.credentials.id} to ${fileName}`);
                return (0, promises_1.writeFile)(fileName, JSON.stringify(sanitizedCredential, null, 2));
            }));
            return {
                count: credentialsToBeExported.length,
                folder: this.credentialExportFolder,
                files: credentialsToBeExported.map((e) => ({
                    id: e.credentials.id,
                    name: path_1.default.join(this.credentialExportFolder, `${e.credentials.name}.json`),
                })),
                missingIds,
            };
        }
        catch (error) {
            throw Error(`Failed to export credentials to work folder: ${error.message}`);
        }
    }
};
exports.SourceControlExportService = SourceControlExportService = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [variables_service_1.VariablesService,
        repositories_1.TagRepository])
], SourceControlExportService);
