// n8n/dist/services/redis/RedisServicePubSubPublisher.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisServicePubSubPublisher = void 0;
const typedi_1 = require("typedi");
const RedisServiceHelper_1 = require("./RedisServiceHelper");
const RedisServiceBaseClasses_1 = require("./RedisServiceBaseClasses");
let RedisServicePubSubPublisher = exports.RedisServicePubSubPublisher = class RedisServicePubSubPublisher extends RedisServiceBaseClasses_1.RedisServiceBaseSender {
    async init(senderId) {
        await super.init('publisher');
        this.setSenderId(senderId);
    }
    async publish(channel, message) {
        var _a;
        if (!this.redisClient) {
            await this.init();
        }
        await ((_a = this.redisClient) === null || _a === void 0 ? void 0 : _a.publish(channel, message));
    }
    async publishToEventLog(message) {
        await this.publish(RedisServiceHelper_1.EVENT_BUS_REDIS_CHANNEL, message.toString());
    }
    async publishToCommandChannel(message) {
        await this.publish(RedisServiceHelper_1.COMMAND_REDIS_CHANNEL, JSON.stringify(message));
    }
    async publishToWorkerChannel(message) {
        await this.publish(RedisServiceHelper_1.WORKER_RESPONSE_REDIS_CHANNEL, JSON.stringify(message));
    }
};
exports.RedisServicePubSubPublisher = RedisServicePubSubPublisher = __decorate([
    (0, typedi_1.Service)()
], RedisServicePubSubPublisher);
