// n8n/dist/services/redis/RedisServiceStreamProducer.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisServiceStreamProducer = void 0;
const typedi_1 = require("typedi");
const RedisServiceHelper_1 = require("./RedisServiceHelper");
const RedisServiceBaseClasses_1 = require("./RedisServiceBaseClasses");
let RedisServiceStreamProducer = exports.RedisServiceStreamProducer = class RedisServiceStreamProducer extends RedisServiceBaseClasses_1.RedisServiceBaseSender {
    async init(senderId) {
        await super.init('producer');
        this.setSenderId(senderId);
    }
    async add(streamName, values) {
        var _a;
        await ((_a = this.redisClient) === null || _a === void 0 ? void 0 : _a.xadd(streamName, '*', 'senderId', this.senderId, ...values));
    }
    async addToEventStream(message) {
        await this.add(RedisServiceHelper_1.EVENT_BUS_REDIS_STREAM, [
            'message',
            message.eventName,
            'event',
            message.toString(),
        ]);
    }
    async addToCommandChannel(message) {
        await this.add(RedisServiceHelper_1.COMMAND_REDIS_STREAM, ['command', JSON.stringify(message)]);
    }
    async addToWorkerChannel(message) {
        await this.add(RedisServiceHelper_1.WORKER_RESPONSE_REDIS_STREAM, ['response', JSON.stringify(message)]);
    }
};
exports.RedisServiceStreamProducer = RedisServiceStreamProducer = __decorate([
    (0, typedi_1.Service)()
], RedisServiceStreamProducer);
