// n8n/dist/services/redis.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedisService = void 0;
const typedi_1 = require("typedi");
const RedisServicePubSubSubscriber_1 = require("./redis/RedisServicePubSubSubscriber");
const RedisServicePubSubPublisher_1 = require("./redis/RedisServicePubSubPublisher");
const RedisServiceListReceiver_1 = require("./redis/RedisServiceListReceiver");
const RedisServiceListSender_1 = require("./redis/RedisServiceListSender");
const RedisServiceStreamConsumer_1 = require("./redis/RedisServiceStreamConsumer");
const RedisServiceStreamProducer_1 = require("./redis/RedisServiceStreamProducer");
let RedisService = exports.RedisService = class RedisService {
    constructor(redisServicePubSubSubscriber, redisServicePubSubPublisher, redisServiceListReceiver, redisServiceListSender, redisServiceStreamConsumer, redisServiceStreamProducer) {
        this.redisServicePubSubSubscriber = redisServicePubSubSubscriber;
        this.redisServicePubSubPublisher = redisServicePubSubPublisher;
        this.redisServiceListReceiver = redisServiceListReceiver;
        this.redisServiceListSender = redisServiceListSender;
        this.redisServiceStreamConsumer = redisServiceStreamConsumer;
        this.redisServiceStreamProducer = redisServiceStreamProducer;
    }
    async getPubSubSubscriber() {
        await this.redisServicePubSubSubscriber.init();
        return this.redisServicePubSubSubscriber;
    }
    async getPubSubPublisher() {
        await this.redisServicePubSubPublisher.init();
        return this.redisServicePubSubPublisher;
    }
    async getListSender() {
        await this.redisServiceListSender.init();
        return this.redisServiceListSender;
    }
    async getListReceiver() {
        await this.redisServiceListReceiver.init();
        return this.redisServiceListReceiver;
    }
    async getStreamProducer() {
        await this.redisServiceStreamProducer.init();
        return this.redisServiceStreamProducer;
    }
    async getStreamConsumer() {
        await this.redisServiceStreamConsumer.init();
        return this.redisServiceStreamConsumer;
    }
};
exports.RedisService = RedisService = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [RedisServicePubSubSubscriber_1.RedisServicePubSubSubscriber,
        RedisServicePubSubPublisher_1.RedisServicePubSubPublisher,
        RedisServiceListReceiver_1.RedisServiceListReceiver,
        RedisServiceListSender_1.RedisServiceListSender,
        RedisServiceStreamConsumer_1.RedisServiceStreamConsumer,
        RedisServiceStreamProducer_1.RedisServiceStreamProducer])
], RedisService);
