// n8n/dist/AbstractServer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractServer = void 0;
const typedi_1 = require("typedi");
const promises_1 = require("fs/promises");
const express_1 = __importDefault(require("express"));
const compression_1 = __importDefault(require("compression"));
const config_1 = __importDefault(require("./config"));
const constants_1 = require("./constants");
const ActiveWorkflowRunner_1 = require("./ActiveWorkflowRunner");
const Db = __importStar(require("./Db"));
const ExternalHooks_1 = require("./ExternalHooks");
const ResponseHelper_1 = require("./ResponseHelper");
const middlewares_1 = require("./middlewares");
const TestWebhooks_1 = require("./TestWebhooks");
const WaitingWebhooks_1 = require("./WaitingWebhooks");
const WebhookHelpers_1 = require("./WebhookHelpers");
const redis_service_1 = require("./services/redis.service");
const n8n_workflow_1 = require("n8n-workflow");
const eventbus_1 = require("./eventbus");
const Helpers_1 = require("./eventbus/EventMessageClasses/Helpers");
const RedisServiceHelper_1 = require("./services/redis/RedisServiceHelper");
const generators_1 = require("./databases/utils/generators");
class AbstractServer {
    constructor(instanceType = 'main') {
        this.instanceId = '';
        this.webhooksEnabled = true;
        this.testWebhooksEnabled = false;
        this.app = (0, express_1.default)();
        this.app.disable('x-powered-by');
        this.protocol = config_1.default.getEnv('protocol');
        this.sslKey = config_1.default.getEnv('ssl_key');
        this.sslCert = config_1.default.getEnv('ssl_cert');
        this.timezone = config_1.default.getEnv('generic.timezone');
        this.restEndpoint = config_1.default.getEnv('endpoints.rest');
        this.endpointWebhook = config_1.default.getEnv('endpoints.webhook');
        this.endpointWebhookTest = config_1.default.getEnv('endpoints.webhookTest');
        this.endpointWebhookWaiting = config_1.default.getEnv('endpoints.webhookWaiting');
        this.uniqueInstanceId = (0, generators_1.generateHostInstanceId)(instanceType);
    }
    async configure() {
    }
    async setupErrorHandlers() {
        const { app } = this;
        const { Handlers: { requestHandler, errorHandler }, } = await Promise.resolve().then(() => __importStar(require('@sentry/node')));
        app.use(requestHandler());
        app.use(errorHandler());
    }
    setupCommonMiddlewares() {
        this.app.use((0, compression_1.default)());
        this.app.use(middlewares_1.rawBodyReader);
    }
    setupDevMiddlewares() {
        this.app.use(middlewares_1.corsMiddleware);
    }
    setupPushServer() { }
    async setupHealthCheck() {
        this.app.get('/healthz', async (req, res) => {
            res.send({ status: 'ok' });
        });
        const { connectionState } = Db;
        this.app.use((req, res, next) => {
            if (connectionState.connected) {
                if (connectionState.migrated)
                    next();
                else
                    res.send('n8n is starting up. Please wait');
            }
            else
                (0, ResponseHelper_1.sendErrorResponse)(res, new ResponseHelper_1.ServiceUnavailableError('Database is not ready!'));
        });
        if (config_1.default.getEnv('executions.mode') === 'queue') {
            await this.setupRedis();
        }
    }
    async setupRedis() {
        const redisService = typedi_1.Container.get(redis_service_1.RedisService);
        const redisSubscriber = await redisService.getPubSubSubscriber();
        await redisSubscriber.subscribeToEventLog();
        await redisSubscriber.subscribeToWorkerResponseChannel();
        redisSubscriber.addMessageHandler('AbstractServerReceiver', async (channel, message) => {
            if (channel === RedisServiceHelper_1.EVENT_BUS_REDIS_CHANNEL) {
                const eventData = (0, n8n_workflow_1.jsonParse)(message);
                if (eventData) {
                    const eventMessage = (0, Helpers_1.getEventMessageObjectByType)(eventData);
                    if (eventMessage) {
                        await eventbus_1.eventBus.send(eventMessage);
                    }
                }
            }
            else if (channel === RedisServiceHelper_1.WORKER_RESPONSE_REDIS_CHANNEL) {
                const workerResponse = (0, n8n_workflow_1.jsonParse)(message);
                if (workerResponse) {
                    console.log('Received worker response', workerResponse);
                }
            }
        });
    }
    async init() {
        const { app, protocol, sslKey, sslCert } = this;
        if (protocol === 'https' && sslKey && sslCert) {
            const https = await Promise.resolve().then(() => __importStar(require('https')));
            this.server = https.createServer({
                key: await (0, promises_1.readFile)(this.sslKey, 'utf8'),
                cert: await (0, promises_1.readFile)(this.sslCert, 'utf8'),
            }, app);
        }
        else {
            const http = await Promise.resolve().then(() => __importStar(require('http')));
            this.server = http.createServer(app);
        }
        const PORT = config_1.default.getEnv('port');
        const ADDRESS = config_1.default.getEnv('listen_address');
        this.server.on('error', (error) => {
            if (error.code === 'EADDRINUSE') {
                console.log(`n8n's port ${PORT} is already in use. Do you have another instance of n8n running already?`);
                process.exit(1);
            }
        });
        await new Promise((resolve) => this.server.listen(PORT, ADDRESS, () => resolve()));
        this.externalHooks = typedi_1.Container.get(ExternalHooks_1.ExternalHooks);
        this.activeWorkflowRunner = typedi_1.Container.get(ActiveWorkflowRunner_1.ActiveWorkflowRunner);
        await this.setupHealthCheck();
        console.log(`n8n ready on ${ADDRESS}, port ${PORT}`);
    }
    async start() {
        if (!constants_1.inTest) {
            await this.setupErrorHandlers();
            this.setupPushServer();
        }
        this.setupCommonMiddlewares();
        if (this.webhooksEnabled) {
            this.app.all(`/${this.endpointWebhook}/:path(*)`, (0, WebhookHelpers_1.webhookRequestHandler)(typedi_1.Container.get(ActiveWorkflowRunner_1.ActiveWorkflowRunner)));
            this.app.all(`/${this.endpointWebhookWaiting}/:path/:suffix?`, (0, WebhookHelpers_1.webhookRequestHandler)(typedi_1.Container.get(WaitingWebhooks_1.WaitingWebhooks)));
        }
        if (this.testWebhooksEnabled) {
            const testWebhooks = typedi_1.Container.get(TestWebhooks_1.TestWebhooks);
            this.app.all(`/${this.endpointWebhookTest}/:path(*)`, (0, WebhookHelpers_1.webhookRequestHandler)(testWebhooks));
            this.app.delete(`/${this.restEndpoint}/test-webhook/:id`, (0, ResponseHelper_1.send)(async (req) => testWebhooks.cancelTestWebhook(req.params.id)));
        }
        if (constants_1.inDevelopment) {
            this.setupDevMiddlewares();
        }
        this.app.use(middlewares_1.bodyParser);
        await this.configure();
        if (!constants_1.inTest) {
            console.log(`Version: ${constants_1.N8N_VERSION}`);
            const defaultLocale = config_1.default.getEnv('defaultLocale');
            if (defaultLocale !== 'en') {
                console.log(`Locale: ${defaultLocale}`);
            }
            await this.externalHooks.run('n8n.ready', [this, config_1.default]);
        }
    }
}
exports.AbstractServer = AbstractServer;
