// n8n/dist/ActiveExecutions.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActiveExecutions = void 0;
const typedi_1 = require("typedi");
const n8n_workflow_1 = require("n8n-workflow");
const utils_1 = require("./utils");
const repositories_1 = require("./databases/repositories");
let ActiveExecutions = exports.ActiveExecutions = class ActiveExecutions {
    constructor() {
        this.activeExecutions = {};
    }
    async add(executionData, process, executionId) {
        let executionStatus = executionId ? 'running' : 'new';
        if (executionId === undefined) {
            const fullExecutionData = {
                data: executionData.executionData,
                mode: executionData.executionMode,
                finished: false,
                startedAt: new Date(),
                workflowData: executionData.workflowData,
                status: executionStatus,
            };
            if (executionData.retryOf !== undefined) {
                fullExecutionData.retryOf = executionData.retryOf.toString();
            }
            const workflowId = executionData.workflowData.id;
            if (workflowId !== undefined && (0, utils_1.isWorkflowIdValid)(workflowId)) {
                fullExecutionData.workflowId = workflowId;
            }
            const executionResult = await typedi_1.Container.get(repositories_1.ExecutionRepository).createNewExecution(fullExecutionData);
            executionId = executionResult.id;
            if (executionId === undefined) {
                throw new Error('There was an issue assigning an execution id to the execution');
            }
            executionStatus = 'running';
        }
        else {
            const execution = {
                id: executionId,
                data: executionData.executionData,
                waitTill: null,
                status: executionStatus,
            };
            await typedi_1.Container.get(repositories_1.ExecutionRepository).updateExistingExecution(executionId, execution);
        }
        this.activeExecutions[executionId] = {
            executionData,
            process,
            startedAt: new Date(),
            postExecutePromises: [],
            status: executionStatus,
        };
        return executionId;
    }
    attachWorkflowExecution(executionId, workflowExecution) {
        if (this.activeExecutions[executionId] === undefined) {
            throw new Error(`No active execution with id "${executionId}" got found to attach to workflowExecution to!`);
        }
        this.activeExecutions[executionId].workflowExecution = workflowExecution;
    }
    attachResponsePromise(executionId, responsePromise) {
        if (this.activeExecutions[executionId] === undefined) {
            throw new Error(`No active execution with id "${executionId}" got found to attach to workflowExecution to!`);
        }
        this.activeExecutions[executionId].responsePromise = responsePromise;
    }
    resolveResponsePromise(executionId, response) {
        var _a;
        if (this.activeExecutions[executionId] === undefined) {
            return;
        }
        (_a = this.activeExecutions[executionId].responsePromise) === null || _a === void 0 ? void 0 : _a.resolve(response);
    }
    remove(executionId, fullRunData) {
        if (this.activeExecutions[executionId] === undefined) {
            return;
        }
        for (const promise of this.activeExecutions[executionId].postExecutePromises) {
            promise.resolve(fullRunData);
        }
        delete this.activeExecutions[executionId];
    }
    async stopExecution(executionId, timeout) {
        if (this.activeExecutions[executionId] === undefined) {
            return;
        }
        if (this.activeExecutions[executionId].process !== undefined) {
            if (this.activeExecutions[executionId].process.connected) {
                setTimeout(() => {
                    this.activeExecutions[executionId].process.send({
                        type: timeout || 'stopExecution',
                    });
                }, 1);
            }
        }
        else {
            this.activeExecutions[executionId].workflowExecution.cancel();
        }
        return this.getPostExecutePromise(executionId);
    }
    async getPostExecutePromise(executionId) {
        const waitPromise = await (0, n8n_workflow_1.createDeferredPromise)();
        if (this.activeExecutions[executionId] === undefined) {
            throw new Error(`There is no active execution with id "${executionId}".`);
        }
        this.activeExecutions[executionId].postExecutePromises.push(waitPromise);
        return waitPromise.promise();
    }
    getActiveExecutions() {
        const returnData = [];
        let data;
        for (const id of Object.keys(this.activeExecutions)) {
            data = this.activeExecutions[id];
            returnData.push({
                id,
                retryOf: data.executionData.retryOf,
                startedAt: data.startedAt,
                mode: data.executionData.executionMode,
                workflowId: data.executionData.workflowData.id,
                status: data.status,
            });
        }
        return returnData;
    }
    async setStatus(executionId, status) {
        if (this.activeExecutions[executionId] === undefined) {
            n8n_workflow_1.LoggerProxy.debug(`There is no active execution with id "${executionId}", can't update status to ${status}.`);
            return;
        }
        this.activeExecutions[executionId].status = status;
    }
    getStatus(executionId) {
        if (this.activeExecutions[executionId] === undefined) {
            return 'unknown';
        }
        return this.activeExecutions[executionId].status;
    }
};
exports.ActiveExecutions = ActiveExecutions = __decorate([
    (0, typedi_1.Service)()
], ActiveExecutions);
